lib/src/annotate.rs

```diff
@@ -0,0 +1,397 @@
+/// local_line_map to say that "Commit B line 6 goes to line 7 of the original
+/// file". We repeat this for all lines in common in the two commits. For 2
+/// identical files, we bulk replace all mappings from commit A to commit B in
+/// local_line_map
+fn process_file_ids(
+    results: &mut PartialResults,
+    store: &Store,
+    file_name: &RepoPath,
+    current_file_id: &FileId,
+    current_commit_id: &CommitId,
+    parent_file_id: &FileId,
+    parent_commit_id: &CommitId,
+) -> BackendResult<()> {
+    results.load_file_into_cache(store, current_commit_id, file_name, current_file_id)?;
+    results.load_file_into_cache(store, parent_commit_id, file_name, parent_file_id)?;
+
+    let current_contents = results.file_cache.get(current_commit_id).unwrap();
+    let parent_contents = results.file_cache.get(parent_commit_id).unwrap();
+
+    let same_lines = get_same_line_map(current_contents, parent_contents);
+    for (current_line_no, parent_line_no) in same_lines {
+        results.forward_to_new_commit(
+            current_commit_id,
+            current_line_no,
+            parent_commit_id,
+            parent_line_no,
+        );
+    }
+    Ok(())
+}
+
+/// For two files, get a map of all lines in common (e.g. line 8 maps to line 9)
+fn get_same_line_map(current_contents: &[u8], parent_contents: &[u8]) -> HashMap<usize, usize> {
+    let mut result_map = HashMap::new();
+    let inputs = vec![current_contents, parent_contents];
+    let diff = Diff::by_line(&inputs);
+    let mut current_line_counter: usize = 0;
+    let mut parent_line_counter: usize = 0;
+    for hunk in diff.hunks() {
+        match hunk {
+            DiffHunk::Matching(common_string) => {
+                for _ in common_string.lines() {
+                    result_map.insert(current_line_counter, parent_line_counter);
+                    current_line_counter += 1;
+                    parent_line_counter += 1;
+                }
+            }
+            DiffHunk::Different(outputs) => {
+                let current_output = outputs[0];
+                let parent_output = outputs[1];
+                if !current_output.is_empty() {
+                    for _ in current_output.lines() {
+                        current_line_counter += 1;
+                    }
+                }
+                if !parent_output.is_empty() {
+                    for _ in parent_output.lines() {
+                        parent_line_counter += 1;
+                    }
+                }
+            }
+        }
+    }
+
+    result_map
+}
+
+fn get_file_id(
+    commit_id: &CommitId,
+    tree: &MergedTree,
+    file_name: &RepoPath,
+) -> Result<Option<FileId>, AnnotateError> {
+    let file_value = tree
+        .path_value(file_name)?
+        .into_resolved()
+        .map_err(|_| {
+            AnnotateError::Conflicted(file_name.to_internal_dir_string(), commit_id.hex())
+        })?
+        .unwrap();
```

Comment:
Maybe better to use materialized conflict instead of erroring out?
See `conflicts::materialize_tree_value()` and its callers.

Reply:
I took a look at this API when implementing. I think this is interesting but not exactly what we want because it includes file dressing for diffs that return false positives. Meaning there is '-' and '+' characters that interfere with the diffing algorithm. 
I think in a later fix when we support conflicts we can annotate the conflicted lines in a smarter way

Reply:
Yeah, I agree about that. I just mean materialized conflict will be okay as a temporary solution, and should be better than disabling annotation at all.

Reply:
I implemented conflicts now according to your suggestions. It now renders the file conflict dressings as added in the current commit